engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala

```diff
@@ -26,8 +26,8 @@ class ExecuteJob(
 
   /** @inheritdoc */
   override def run(implicit ctx: RuntimeContext): Unit = {
-    ctx.locking.acquireContextLock(contextId)
     ctx.locking.acquireReadCompilationLock()
+    ctx.locking.acquireContextLock(contextId)
```

Comment:
If you want to impose a _"hierarchy of locks"_ (one way to break the [four necessary deadlock conditions](http://wiki.apidesign.org/wiki/Deadlock_conditions), then I suggest to add additional: `isHeldByCurrentThread()` checks into `ReentrantLocking`. In this case it'd be:
In `acquireReadCompilation` verify whether `contextsLock.isEmpty()` - if not, throw an error indicating _"violation of locks hierarchy"_. Etc.

Reply:
There could be multiple jobs each having acquired `readCompilation` lock, and having a different `contextId` which would be fine from the execution point of view, I think, rendering that kind of check invalid?